e2etest/bgptests/metrics.go

```diff
@@ -0,0 +1,760 @@
+		for _, pod := range speakers {
+			ginkgo.By(fmt.Sprintf("checking pod %s", pod.Name))
+			Eventually(func() error {
+				podMetrics, err := metrics.ForPod(controllerPod, pod, metallb.Namespace)
+				framework.ExpectNoError(err)
+				err = metrics.ValidateGaugeValue(1, "metallb_k8s_client_config_stale_bool", map[string]string{}, podMetrics)
+				if err != nil {
+					return err
+				}
+				err = metrics.ValidateOnPrometheus(promPod, fmt.Sprintf(`metallb_k8s_client_config_stale_bool{pod="%s"} == 1`, pod.Name), metrics.There)
+				if err != nil {
+					return err
+				}
+				return nil
+			}, time.Minute, 1*time.Second).ShouldNot(HaveOccurred(), "on pod", pod.Name)
+		}
+
+		resources.BFDProfiles = []metallbv1beta1.BFDProfile{bfdProfile}
+		err = ConfigUpdater.Update(resources)
+		framework.ExpectNoError(err)
+		for _, pod := range allPods {
+			ginkgo.By(fmt.Sprintf("checking pod %s", pod.Name))
+			Eventually(func() error {
+				podMetrics, err := metrics.ForPod(controllerPod, pod, metallb.Namespace)
+				framework.ExpectNoError(err)
+				err = metrics.ValidateGaugeValue(0, "metallb_k8s_client_config_stale_bool", map[string]string{}, podMetrics)
+				if err != nil {
+					return err
+				}
+				err = metrics.ValidateOnPrometheus(promPod, fmt.Sprintf(`metallb_k8s_client_config_stale_bool{pod="%s"} == 0`, pod.Name), metrics.There)
+				if err != nil {
+					return err
+				}
+				// we don't know how many events we are processing
+				err = metrics.ValidateCounterValue(metrics.GreaterThan(0), "metallb_k8s_client_updates_total", map[string]string{}, podMetrics)
+				if err != nil {
+					return err
+				}
+				err = metrics.ValidateOnPrometheus(promPod, fmt.Sprintf(`metallb_k8s_client_updates_total{pod="%s"} > 0`, pod.Name), metrics.There)
+				if err != nil {
+					return err
+				}
+				err = metrics.ValidateGaugeValue(1, "metallb_k8s_client_config_loaded_bool", map[string]string{}, podMetrics)
+				if err != nil {
+					return err
+				}
+				err = metrics.ValidateOnPrometheus(promPod, fmt.Sprintf(`metallb_k8s_client_config_loaded_bool{pod="%s"} == 1`, pod.Name), metrics.There)
+				if err != nil {
+					return err
+				}
+				return nil
+			}, time.Minute, 5*time.Second).ShouldNot(HaveOccurred(), "on pod", pod.Name)
+		}
+	})
+})
+
+type peerPrometheus struct {
+	labelsForQueryBGP string
+	labelsBGP         map[string]string
+	labelsForQueryBFD string
+	labelsBFD         map[string]string
+	noEcho            bool
+}
+
+func labelsForPeers(peers []*frrcontainer.FRR, ipFamily ipfamily.Family) []peerPrometheus {
+	res := make([]peerPrometheus, 0)
+	for _, c := range peers {
+		address := c.Ipv4
+		if ipFamily == ipfamily.IPv6 {
+			address = c.Ipv6
+		}
+		peerAddr := address + fmt.Sprintf(":%d", c.RouterConfig.BGPPort)
+
+		labelsBGP := map[string]string{"peer": peerAddr}
+		labelsForQueryBGP := fmt.Sprintf(`peer="%s"`, peerAddr)
+		labelsBFD := map[string]string{"peer": address}
+		labelsForQueryBFD := fmt.Sprintf(`peer="%s"`, address)
```

Comment:
I think we should have a `"vrf": default`, `vrf="default"` here, is this omitted because of native?

Reply:
I wanted to validate that if a user is not using vrfs, the old label selection would still work, so for the default vrf we select without labels. Let me add a comment for that so our future selves will thank.